src/ray/object_manager/pull_manager.cc

```diff
@@ -314,61 +314,50 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
-  //    or remote node that spilled the object.
-  // 2. Also, if we use multi-node file spilling, the restoration will be
-  //    confirmed by a object location subscription, so we should pull first
-  //    before requesting for object restoration.
+  // Fall back to pulling from a remote node. If the object is spilled on the local
+  // disk of the remote node, it will be restored by PushManager prior to pushing.
   bool did_pull = PullFromRandomLocation(object_id);
   if (did_pull) {
-    // New object locations were found, so begin trying to pull from a
-    // client.
     UpdateRetryTimer(request);
     return;
   }
 
-  // If we cannot pull, it means all objects have been evicted, so try restoring objects
-  // from the external storage. If the object was spilled on the current node, the
-  // callback will restore the object from the local the disk.
-  // Otherwise, it will send a request to a remote node that spilled the object.
-  // If external storage is a distributed storage, we always try restoring from it without
-  // sending RPCs.
-  if (!request.spilled_url.empty()) {
-    const auto spilled_node_id = request.spilled_node_id;
-    restore_spilled_object_(
-        object_id, request.spilled_url, spilled_node_id,
-        [this, object_id, spilled_node_id](const ray::Status &status) {
-          if (!status.ok()) {
-            const auto node_id_with_issue =
-                spilled_node_id.IsNil() ? self_node_id_ : spilled_node_id;
-            RAY_LOG(WARNING)
-                << "Object restoration failed and the object could "
-                   "not be "
-                   "found on any other nodes. This can happen if the location where the "
-                   "object was spilled is unreachable. This job may hang if the object "
-                   "is permanently unreachable. "
-                   "Please check the log of node of id: "
-                << node_id_with_issue << " Object id: " << object_id;
-          }
-        });
-    // We shouldn't update the timer here because restoration takes some time, and since
-    // we retry pull requests with exponential backoff, the delay could be large.
+  // If we can restore directly from this raylet, then try to do so.
+  bool can_restore_directly =
+      !request.spilled_url.empty() &&
+      (request.spilled_node_id.IsNil() || request.spilled_node_id == self_node_id_);
+  if (can_restore_directly) {
+    UpdateRetryTimer(request);
+    restore_spilled_object_(object_id, request.spilled_url,
+                            [this, object_id](const ray::Status &status) {
+                              if (!status.ok()) {
+                                RAY_LOG(ERROR) << "Object restore for " << object_id.Hex()
+                                               << " failed, will retry later: " << status;
+                              }
+                            });
+    return;
   }
+
+  // TODO(ekl) should we more directly mark the object as lost in this case?
```

Comment:
Hmm @stephanie-wang what should we do in this case? (this is an existing problem I guess). It seems we will be rapidly re-polling the object during Tick().

Reply:
Isn't it something that can happen before the spilled object URL / location is updated? (e.g., the object location is not available yet). If so, don't we just need to ignore this case?

Reply:
Hmm yeah it seems like we'll want to mark the object as failed. Or we can let the owner handle it now that we have the OBOD.

Reply:
Hmm if we publish the URL update at the same time we remove the location, then I don't see how this could happen. The only case would be for node failure... I guess currently task dep manager will mark it as failed and cancel the pull at a higher level hopefully